### PR TITLE
Passive AntiCheat

### PR DESCRIPTION
**Pull request by sveN295**
**Our train of thought:**

As a game over 10 years old and still popular, World of Warcraft: Wrath of the Lich King servers suffer greatly from cheating players.
TrinityCore does not offer an effective solution for this.
Most approaches are (sometimes for good reasons) closed source.
The problem with server-side anti-cheat software is less the detection of cheaters, but the collateral damage to innocent players through false-positive detection.

Due to this, we believe that server-side measures should not work automatically, but should support the admin to identify cheaters.
Therefore we "only" provide methods for database logging. Its up to the admin/users to take actions based on the collected data.

There are a number of factors that (can) have a negative impact on cheat detection:
- Quality of maps
- Latency
- Server load

We got a hand on the 3 most commonly used cheat-tools (WoWAdminPanel, WoW Hitchhiker, Gagarin) and worked on detecting the most harmful hacks these tools offer:

- Teleporting
- Flyhacking (+ fake Flyhacking)
- Speedhacking
- Climbhacking
- Waterwalking
- Anti-Swim hacking
- Jumphacking
- Ignore control effect hacks
- Gravityhacking
- Anti Knockback hack

Aside from some hacks where you want to identify every single violation (like teleport hack), most hacks are permanently or for longer periodes active (like speedhack).
World of Warcraft has a ton of spells and mechanics that leads to false positive detections. Warriors charge, Deathknights pull, Mages Blink, Rogues Shadowstep and so on are known for giving false positive detections.

We looked at other approaches but decided against hooking into spell effects or write strange exceptions and huge tolerance numbers to avoid edge cases.
Our solution here is that a innocent player only triggers movement related checks in edge cases every now and then. A cheater will continuously trigger detections.
Therefore we generate a time frame of (currently) 5 seconds where the player has to (currently) hit 3 cheat detections to generate a report. 

But as everyone that ever worked on a wow server emulator, we are now at a point where additional tests to collect more data are too much for just 2 people.

**Changes proposed**:

- Passive AntiCheat System for different hack types

**Target branch(es)**: 335

**Issues addressed**: Fixes #

**Tests performed**: (Does it build, tested in-game, etc)
Builds and tested in-game

**Known issues and TODO list**:

- [ ] Collect user data

### CODE REVIEW

File: src/server/game/NemesisAntiCheat/AntiCheatMgr.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#ifndef SC_ACMGR_H
+#define SC_ACMGR_H
+
+#include "Common.h"
+#include "SharedDefines.h"
+#include "ScriptMgr.h"
+#include "Chat.h"
+#include "Player.h"
+#include "World.h"
+#include "WorldSession.h"
```
Comment (jackpoz) [inline src/server/game/NemesisAntiCheat/AntiCheatMgr.h]:
```suggestion
```

File: src/server/game/NemesisAntiCheat/AntiCheatMgr.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+class Player;
+class WorldSession;
```
Comment (jackpoz) [inline src/server/game/NemesisAntiCheat/AntiCheatMgr.h]:
```suggestion
```

File: src/server/game/NemesisAntiCheat/AntiCheatMgr.cpp [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+https://github.com/Jinnaix
+With feedback and support from https://github.com/acidmanifesto
+*/
+
+#include "AntiCheatMgr.h"
+#include "DatabaseEnv.h"
+#include "Map.h"
+#include "DBCStores.h"
+#include "Realm.h"
+#include "Log.h"
```
Comment (jackpoz) [inline src/server/game/NemesisAntiCheat/AntiCheatMgr.cpp]:
```suggestion
#include "DatabaseEnv.h"
#include "DBCStores.h"
#include "Log.h"
#include "Map.h"
#include "Realm.h"
#include "WorldSession.h"
```

**Review by jackpoz — CHANGES_REQUESTED** [review]
I left some comments about nopch fix. Headers should be included in .cpp files, with classes forward declared in .h when needed.

**Review by jackpoz — APPROVED** [review]

**Review by acidmanifesto — APPROVED** [review]
Just leaving some background info in regards to this anticheat.

 I think it is a great approach for a alternative anticheat.

 SveN is one of a few people who has help me personally with numerous detections and code style assistance dealing with the current passive anticheat which was a very huge and impressive undertaking, with out his assistance the current passive anticheat would not be at the level of propper detection it is currently.

His approach in detection is different and does provide amazing results as well from first hand experience.

I see his anticheat as something for advance use and made for manual moderation without in game spam notifications.

I do strongly consider this as a great alternative for users who wish for a alternative anticheat detection approach from the current.

I firmly his detections are on par with the current passive anticheat if not a bit different in its detection approach in ensuring positive detection.

**Comment by jackpoz** [comment]
All the branches on this repo have a short description at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki . What would be a good description for this one ?

**Comment by sveN295** [comment]
Short Description:

> Passive Anticheat Detection for various hacks (Speedhack, Flyhack, Teleport, Jumphack, Climbhack, ZAxisHack, Waterwalk, Antiswim, IgnoreControl) with detailed Database logging

**Comment by sveN295** [comment]
I also prepared a longer description for a wiki page:

**Features**

•	Passive hack detection for multiple hacks tested with different paid hacking tools
•	Detects hacks by looking at movement flags, opcodes and calculating values with server-side knowledge
•	Clean console and database logging with a custom view to identify frequent offenders
•	System in place to reduce false positives

**Types of hacks detected**

•	Speedhack: Increase of movement speed.
•	Flyhack: Flying without flying mounts.
•	Teleporthack: Teleport to a selected location.
•	Jumphack: Jump again before you land on the ground.
•	Gravityhack: Lower gravity allows higher jumps.
•	Climbhack: Climb impossible angles. Can detect certain types of No Fall Damage Hacks as well.
•	ZAxisHack: Disable falling so you can run in the air.
•	Waterwalk: Walk on water.
•	Antiswim: Walk under water.
•	Ignore Control: Ignore roots.
•	Anti-Knockback: Ignore knockbacks.

**Premise**

Wrath of the Lich King is still a very popular game that attracts a lot of players on private servers. This also means that there is a huge market for cheat tools.
TrinityCore only has basic functionality to detect fake movement flags. This detection is entirely side-stepped by commen paid hack tools. Anticheat modules that can detect these hacks are mostly closed-source and paid.
The problem with server-side anti-cheat software is less the detection of cheaters, but the collateral damage to innocent players through false-positive detection.
Due to this, we believe that server-side measures should not work automatically, but should support the admin to identify cheaters.
Therefore we "only" provide methods for database logging. Its up to the admin/users to take actions based on the collected data.
There are a number of factors that (can) have a negative impact on cheat detection:

•	Quality of maps
•	Latency
•	Server load

We got our hands on 3 frequently used cheat tools (WoWAdminPanel, WoW Hitchhiker, Gagarin) and worked on detecting the most harmful hacks these tools offer.

**Hack Detection**

There are hacks where you want to catch every violation, for example teleport hacks. Then there are also hacks where detection is about continous violation of the rules, for example speed hack, where someone is travelling faster than allowed for extended periods of time.
For the second type, we were able to reduce the amount of false positives by allowing every player a certain amount of detections (default 3) in a certain amount of time (default 5 seconds). We believe that legit players will have a hard time accidentally triggering the anti cheat detection  this often while our test have shown that cheaters will trigger it still.
We also tried to find better solutions than to write exceptions for specific spells or areas.
When a cheater is detected, we output a warning in the console. There will also be an incident report in the „nemesis_anticheat“ table in the auth database that offers valuable information about the incident:

•	Time of the report
•	Realm ID
•	Account ID and Account Name
•	Character GUID and Character Name
•	Character Class
•	What hack was detected and useful values depending on the type of hack
•	Where the detection took place
•	A gm command to telepor directly to this location

But as everyone that ever worked on a wow server emulator, we are now at a point where additional tests to collect more data are too much for just two people. This is why we decided to make the script available for the community to gather more usage data. We hope to use this feedback to further reduce false positives.

**Config**

```
###################################################################################################
#   Nemesis Anti-Cheat System
#
#   Anticheat.Enable
#      Description: Enables or disables the Anticheat System
#      Default:     0 - (Disabled)
#                   1 - (Enabled)

Anticheat.Enabled = 1

#   False Positive Reduction
#   Hack.Counter        (Int, Default 3)   - How many positive hack detections a player needs to generate a report
#   Reset.Timer    (Int, Default 5000)  - How much time in ms we want to wait until resetting the hack counter
#
#   Example: With default values, a player has to trigger a hack detection 3 times in the span of 5 seconds to generate a cheat report
Anticheat.Hack.Counter = 3
Anticheat.Reset.Timer = 5000

#   Anticheat Subsystems - Enable or Disable as needed
#   Speedhack - cheaters running faster than allowed
#   Tolerance - Percentage increase of allowed speed (Default 1.075 for 7.5% increase)
Anticheat.SpeedHack = 1
Anticheat.SpeedHack.Tolerance = 1.075

#   Flyhack - cheaters flying when not allowed
Anticheat.FlyHack = 1

#   TeleportHack - cheaters teleporting from A to B when not allowed
Anticheat.TeleportHack = 1

#   JumpHack - cheaters using multiple jumps at once, reaching higher areas
Anticheat.JumpHack = 1

#   AntiknockbackHack - cheaters ignoring knockbacks
Anticheat.AntiknockbackHack = 1

#   GravityHack - Manipulating gravity so player can jump higher
Anticheat.GravityHack = 1

#   ClimbHack - cheaters climbing hills they should not be able to
#   MaxSlope - Maximum angle in radians a player can climb (Default 55 degrees converted to radian -> 0.9599)
Anticheat.ClimbHack = 1
Anticheat.ClimbHack.MaxSlope = 0.9599

#   ZAxisHack - cheaters hovering
Anticheat.ZAxisHack = 1

#   WaterwalkHack - cheaters walking on water and can disengange from combat that way
Anticheat.WaterwalkHack = 1

#   AntiswimHack - cheaters walking in water
Anticheat.AntiswimHack = 1

#   IgnoreControlHack - cheaters ignoring control loss effects
Anticheat.IgnoreControlHack = 1

#
###################################################################################################
```

**Comment by sveN295** [comment]
**Screenshots**

anticheat_logging_console_1
_Console Logging_

anticheat_logging_table_1
_Database Logging_

anticheat_logging_table_2
_Custom Database View_

**Comment by acidmanifesto** [comment]
I suggest maybe naming the branch Nemesis Passive Anticheat just to ensure people will know it is a different anticheat from the current.

**Comment by jackpoz** [comment]
PR merged, wiki updated, wiki page added at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki/3.3.5-Nemesis-Anticheat 

Let me know if I should update the wiki or so. I will include this branch in the weekly automatic updates

**Comment by jackpoz** [comment]
Thanks for the new anticheat btw!

**Pull request merged**